### DISCUSSION TITLE
How to validate base64?

### DISCUSSION THREAD

**Discussion by gustavotoyota** [colinhacks/zod#1446] [Q&A] [answered]
Very simple problem, but can't find any info anywhere.

**Comment by jsjoeio** [comment, marked as answer]
Did you ever figure out a solution?

**Reply by gustavotoyota** [reply]
I used the js-base64 library like this:
```ts
import { Base64 } from 'js-base64';

z.object({
  myString: z.string().refine(Base64.isValid),
})
```

**Reply by jsjoeio** [reply]
Amazing! Thank you!

**Reply by brandonwie** [reply]
I tried with a string that is returned from `FileReader`'s `readAsDataURL` which is a base64 encoded string, but it won't pass the validation using `js-base64`. You need to encode with `js-base64` to validate with `Base64.isValid`.
 
For instance,
```typescript
const getBase64 = (
  file: File,
  callback: (result: string | ArrayBuffer | null) => void
) => {
  const reader = new FileReader();
  reader.readAsDataURL(file);
  reader.onload = function () {
    // if you don't encode with Base64, validation with zod on back-end throws an error
    const base64 = Base64.encode(reader.result as string);
    callback(base64);
  };
  reader.onerror = function (error) {
    console.log('Error: ', error);
  };
};
```

**Reply by jsjoeio** [reply]
I ended up doing the same thing!

```ts
z.string().refine(Base64.isValid),
```

**Reply by gustavotoyota** [reply]
@brandonwie I think that's because data URLs are prefixed by the media type, and the base64 part comes after that.

**Comment by mbenedettini** [comment]
For the record, this can now be easily done using `base64()` and without an external library:

```typescript base64.ts
import { z } from "zod";

const validateAndParseBase64 = z
  .string()
  .base64()
  .transform((val: string) => Buffer.from(val, "base64").toString("utf-8"));

console.log(validateAndParseBase64.parse("aGVsbG8gd29ybGQ="));
validateAndParseBase64.parse("should fail"); // will throw
```

**Reply by lwinged** [reply]
Just for information

```
validateAndParseBase64.parse("should"); // ❌ throws: invalid base64 format
validateAndParseBase64.parse("fail");   // ✅ passes: valid base64 format
validateAndParseBase64.parse("test");   // ✅ passes: valid base64 format
```

Explanation:
Although the strings "fail" and "test" may look arbitrary, they are in fact valid base64-encoded strings (their length and character set match the base64 specification, and decoding does not throw an error). The string "should" is not valid base64 and throws an error as expected.

[Base64 Specification (RFC 4648)](https://datatracker.ietf.org/doc/html/rfc4648#section-4)
